refactor pipelines to include contentSources

faberf opened the pull request:
This PR addresses the problem of multiple branches in a pipeline getting in each others way, by changing the use of retrievables to be append only. It is still the job of pipeline configurator to avoid race conditions by making sure that only parts of the retrievable are processed that are not being further updated. This is made easier via configurable contentSources which allow the configurator to specify which content should be used for an operator using a white list of other operators. This description is not very complete and the code needs to be cleaned up. Therefore this PR is a draft.

- [ ] Add documentation in wiki
- [x] Clean up code
- [x] Remove Passthrough in favor of allowing stages to be configured without an operator

sauterl commented on vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/transform/ContentMergingTransformer.kt:
```diff
@@ -0,0 +1,54 @@
+            input.toFlow(scope).collect { retrievable: Retrievable ->
+                val mergedContent = StringBuilder()
+                contentFields.forEach { fieldName ->
+                    retrievable.getContent(fieldName)?.let { content ->
```
I see two problems with this:

1. I'm not aware of a `getCnotent` on retrievables and
2. Wouldn't you require some checks to see wether this is (reasonable) text ?

Maybe there are other changes missing?

v0idness replied:
Thank you for this feedback, you are entirely correct; this is at a draft stage and hence all this will be addressed.

faberf commented:
Hey, this PR is ready to review. We will add the documentation to the wiki once its merged.

lucaro commented on vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/transform/TemplateTextTransformer.kt:
```diff
@@ -0,0 +1,73 @@
+        val template = context[name, "template"] ?: throw IllegalArgumentException("The template text transformer requires a template.")
+        val regex = "\\$\\{([^}]+)\\}".toRegex()
+        val contentFields = regex.findAll(template).map { it.groupValues[1] }.toList()
+        val defaultValue = context[name, "defaultValue"] ?: ""
```
magic values should be defined as constants

ppanopticon replied:
Especially if they're RegEx, which incur a cost for compilation.

lucaro commented on vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/transform/TemplateTextTransformer.kt:
```diff
@@ -0,0 +1,73 @@
+                    val fieldContent = StringBuilder()
+                    contentIds?.forEach{ id ->
+                        retrievable.content.find {
+                            it.id == id && it.type == ContentType.TEXT
```
In this case, you probably want to consider both `TEXT` and `STRING` as possible inputs

v0idness replied:
Can a `ContentElement` take a type that isn't defined by `ContentType`, such as `STRING` ?

faberf replied:
Yes, I think luca confused the content type with the descriptor subclasses

lucaro commented on vitrivr-engine-module-fes/src/main/kotlin/org/vitrivr/engine/base/features/external/api/AsrApi.kt:
```diff
@@ -35,6 +36,22 @@ class AsrApi(host: String, model: String, timeoutMs: Long, pollingIntervalMs: Lo
+        return try {
+            this.automatedSpeechRecognitionApi.newBatchedJobApiTasksAutomatedSpeechRecognitionBatchedModelJobsPost(this.model, wrapped).body()
+        } catch (e: Throwable) {
+            JobStatus("unknown", JobState.failed)
```
Why "unknown"? The `Throwable` might know why...

faberf replied:
It's not the status that is unknown, it is the job id that is unknown if there is a failure during job creation.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/features/AbstractBatchedExtractor.kt:
```diff
@@ -36,33 +39,39 @@ abstract class AbstractBatchedExtractor<C : ContentElement<*>, D : Descriptor>(f
+                    }
+                } catch (e: Exception) {
+                    logger.error(e) { "Error during extraction" }
+                    "Error during extraction: $e".let {
```
Why are you reporting the error twice?

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/content/impl/cache/CachedAudioContent.kt:
```diff
@@ -35,6 +36,7 @@ class CachedAudioContent(override val path: Path, override val channels: Short,
             }
             return buffer.asReadOnlyBuffer()
         }
+    override val id: UUID = UUID.randomUUID()
```
I would add this to the constructor. There might be use-cases, where one wants to define this explicitly. You can also use default values in the constructor.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/content/impl/cache/CachedImageContent.kt:
```diff
@@ -16,6 +17,8 @@ import javax.imageio.ImageIO
  */
 class CachedImageContent(override val path: Path, image: BufferedImage) : ImageContent, CachedContent {
 
+    override val id: UUID = UUID.randomUUID()
```
I would add this to the constructor. There might be use-cases, where one wants to define this explicitly.You can also use default values in the constructor.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/content/impl/cache/CachedTextContent.kt:
```diff
@@ -14,6 +15,8 @@ import java.nio.file.StandardOpenOption
  */
 class CachedTextContent(override val path: Path, text: String) : TextContent, CachedContent {
 
+    override val id: UUID = UUID.randomUUID()
```
I would add this to the constructor. There might be use-cases, where one wants to define this explicitly. You can also use default values in the constructor.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/content/impl/memory/InMemoryAudioContent.kt:
```diff
@@ -11,4 +12,6 @@ import java.nio.ShortBuffer
  */
-data class InMemoryAudioContent(override val channels: Short, override val samplingRate: Int, override val content: ShortBuffer) : AudioContent
+data class InMemoryAudioContent(override val channels: Short, override val samplingRate: Int, override val content: ShortBuffer) : AudioContent {
+    override val id: UUID = UUID.randomUUID()
```
I would add this to the constructor. There might be use-cases, where one wants to define this explicitly. You can also use default values in the constructor.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/content/impl/memory/InMemoryImageContent.kt:
```diff
@@ -11,4 +12,6 @@ import java.awt.image.BufferedImage
  */
-data class InMemoryImageContent(override val content: BufferedImage) : ImageContent
+data class InMemoryImageContent(override val content: BufferedImage) : ImageContent {
+    override val id: UUID = UUID.randomUUID()
```
I would add this to the constructor. There might be use-cases, where one wants to define this explicitly. You can also use default values in the constructor.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/content/impl/memory/InMemoryMeshContent.kt:
```diff
@@ -12,4 +13,6 @@ import org.vitrivr.engine.core.model.mesh.Model3D
  */
-data class InMemoryMeshContent(override val content: Model3D) : Model3DContent
+data class InMemoryMeshContent(override val content: Model3D) : Model3DContent {
+    override val id: UUID = UUID.randomUUID()
```
I would add this to the constructor. There might be use-cases, where one wants to define this explicitly. You can also use default values in the constructor.

ppanopticon commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/retrievable/attributes/ContentAuthorAttribute.kt:
```diff
@@ -0,0 +1,39 @@
+import java.util.*
+import kotlin.collections.HashMap
+
+class ContentAuthorAttribute private constructor(
```
**I have a very fundamental question:** Why do we need such a `ContentAuthorAttribute` that maps UUIDs to authors? Why don't we just attach the author property directly to the `ContentElement`? 

This would remove one level indirection and prevents inconsistencies, e.g., if an `Operator` removes content but forgets to update the `ContentAuthorAttribute`.

faberf replied:
I think it is a bad design choice to grow the content classes as we make the pipelines more complex. This complexity should be outsourced in seperate objects in my opinion, with the contents themselves only needing an id. Furthermore, an operator should never remove content if we want to have pipelines that support branching.

faberf replied:
Resolving this after in person discussion.

ppanopticon reviewed (dismissed):
See comments. My fundamental question is, if we really need this indirection introduced by the `ContentAuthorAttribute`, or if we'd rather just move this information to the actual `ContentElement`.

faberf commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/retrievable/attributes/ContentAuthorAttribute.kt:
```diff
@@ -0,0 +1,39 @@
+import kotlin.collections.HashMap
+
+class ContentAuthorAttribute private constructor(
+    private val authorMap: HashMap<UUID, HashSet<String>>,
```
The datatype should be very fast at determining if a content has an author. Iterating through a list might be too slow, was my concern.

faberf commented on vitrivr-engine-core/src/main/kotlin/org/vitrivr/engine/core/model/retrievable/attributes/ContentAuthorAttribute.kt:
```diff
@@ -0,0 +1,39 @@
+        return ContentAuthorAttribute(authorMap, contentMap)
+    }
+
+    fun getAuthors(contentId: UUID): Set<String> {
```
see above.

faberf commented on vitrivr-engine-index/src/main/kotlin/org/vitrivr/engine/index/aggregators/AbstractAggregator.kt:
```diff
@@ -25,10 +27,18 @@ abstract class AbstractAggregator(override val input: Operator<out Retrievable>,
      *  @param scope [CoroutineScope] to use for the [Flow].
      */
     override fun toFlow(scope: CoroutineScope): Flow<Retrievable> = this.input.toFlow(scope).map {
+        val contentSources = context.getProperty(name, "contentSources")?.split(",")?.toSet()
```
I agree. A quick global search for "context.getProperty(" reveals this is currently very inconsistant. For this PR I will define CONTENT_AUTHORS_KEY

faberf commented:
It seems like the two concerns left unchanged in the latest commits are:

- The need for a level of indirection via attributes at all
- The unordered implementation of authors associated with a content (and likewise the unordered implementation of contents associated with an author)

I have made the appropriate comments addressing these issues above.

ppanopticon commented:
What is the state of this PR? Have all issues been addressed?

Currently, there are two merge-conflicts which ought to be resolved. Also, please make sure that all tests run successfully.

faberf commented:
> What is the state of this PR? Have all issues been addressed?
> 
> Currently, there are two merge-conflicts which ought to be resolved. Also, please make sure that all tests run successfully.

Me and raphi will merge this today
